docs/docs/canonicai/chapter3.md

```diff
@@ -0,0 +1,339 @@
+
+
+### **Inheritance**
+**What is inheritance?**
+ It is a mechanism where you can to derive a class from another class for a hierarchy of classes that share a set of attributes and methods.
+ 
+ **Node Inheritance**
+``` 
+node vehicle {
+    has plate_number;
+    
+    can drive {
+        report 'yes';
+    }
+}
+
+node car:vehicle {
+    has plate_number = "RAC001";
+}
+
+node bus:vehicle {
+    has plate_number = "SUB002";
+}
+```
+In the above code snippet. This is a very basic example of inheritance. We created a node named car and bus but since they are related and simply share the same property and node ability we created a node named vehicle and inherit its attributes. Imagine life without inheritance, the code will look like this.
+``` 
+node car {
+    has plate_number = "RAC001";
+    
+    can drive {
+        report 'yes';
+    }
+}
+
+node bus {
+    has plate_number = "SUB002";
+    
+    can drive {
+        report 'yes';
+    }
+}
+```
+In this case it will seem that writing it this way will be more effient, however what if each vehicles have more functions like blow_horn, accelerate, breaks, park and etc or they are more vehicles like truck, scooter, and etc you can imagine how big this code base will be and that's why it's important to have inheritance in our code base and in this program we use inheritance to increase the efficiency of the entire application.
+
+So, lets use what's in the code base and see how we implemented it. ```Do not worry about what each nodes and state works, that will be explained in detailed in the next section, just look at how it is implemented.```
+```
+node cai_state {
+    has name;
+    has prepared_entities = {
+            "name": "Tony Stark",
+            "address": "10880 Malibu Point"
+    };
+
+    # Classify the intent of the question
+    can classify_intent {
+        # NOTE: Hardcode intent for now
+        visitor.predicted_intent = visitor.intent_override;
+    }
+
+    # Extract words of interests from the question
+    can extract_entities {
+        # NOTE: Hardcode entities for now
+        for ent in visitor.entity_override {
+            visitor.extracted_entities[ent] = prepared_entities[ent];
+        }
+    }
+
+    # Generate response
+    can gen_response {
+        # Default response
+        visitor.response =
+            "Sorry I can't handle that just yet! Anything else I can help you with";
+    }
+
+    # Process incoming request through NLU engines
+    can nlu {
+        ::classify_intent;
+        ::extract_entities;
+    }
+
+    # Construct natural language response
+    can nlg {
+        ::gen_response;
+    }
+}
+```
+Above, this is a conversational state node. We use inheritance here, because we will have multiple conversational state nodes like a confirm state, cancel state, collect information state and etc, that have the same functionality but does have different node abilities.
+
+```
+node collect_info:cai_state {
```

Comment:
Let's break this down into a code snippet + explanation for each node. For the first one, collect_info, you can explain why we are override gen_response and why we are **only** override gen_response. For the other ones, the explanation can be brief.

Reply:
Oh you actually are breaking this down later. What's the purpose of this code snippet here?

Reply:
The purpose of the code snippet it to show how we used inheritance in the application. It's to show them there is no big difference between that example vs this code snippet.

Reply:
Ok.